src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+
+KJ_TEST("An empty user bundle") {
+  // We should be able to create an empty user bundle that returns nothing.
+  // Basic resolution of this kind does not require an isolate lock.
+
+  ModuleBundle::BundleBuilder builder;
+  auto bundle = builder.finish();
+
+  // Resolve should return nothing.
+  ResolveContext context = {
+    .type = ResolveContext::Type::BUNDLE,
+    .source = ResolveContext::Source::OTHER,
+    .specifier = "file:///foo"_url,
+    .referrer = ModuleBundle::BundleBuilder::BASE,
+  };
+
+  KJ_ASSERT(bundle->resolve(context) == kj::none);
+}
+
+// ======================================================================================
+
+KJ_TEST("An empty built-in bundle") {
+  // We should be able to create an empty built-in bundle that returns nothing.
+  // Basic resolution of this kind does not require an isolate lock.
+
+  ModuleBundle::BuiltinBuilder builder;
+  auto bundle = builder.finish();
+
+  // Resolve should return nothing.
+  ResolveContext context = {
+    .type = ResolveContext::Type::BUNDLE,
+    .source = ResolveContext::Source::OTHER,
+    .specifier = "file:///foo"_url,
+    .referrer = ModuleBundle::BundleBuilder::BASE,
+  };
+
+  KJ_ASSERT(bundle->resolve(context) == kj::none);
+}
+
+// ======================================================================================
+
+KJ_TEST("A registry with empty bundles") {
+  // We should be able to create a registry with empty bundles that return nothing.
+  // Basic resolution of this kind does not require an isolate lock.
+
+  ResolveObserverImpl observer;
+  ModuleRegistry::Builder registryBuilder(
+      observer,
+      ModuleRegistry::Builder::Options::ALLOW_FALLBACK);
+
+  registryBuilder.add(ModuleBundle::newFallbackBundle(
+      [](const ResolveContext& context) {
+    return kj::none;
+  }));
+
+  ModuleBundle::BundleBuilder bundleBuilder;
+  registryBuilder.add(bundleBuilder.finish());
+
+  ModuleBundle::BuiltinBuilder builtinBuilder;
+  registryBuilder.add(builtinBuilder.finish());
+
+  auto registry = registryBuilder.finish();
+
+  // Resolve should return nothing.
+  ResolveContext context = {
+    .type = ResolveContext::Type::BUNDLE,
+    .source = ResolveContext::Source::OTHER,
+    .specifier = "file:///foo"_url,
+    .referrer = ModuleBundle::BundleBuilder::BASE,
+  };
+
+  KJ_ASSERT(registry->resolve(context) == kj::none);
+  KJ_ASSERT(observer.modules.size() == 1);
+  KJ_ASSERT(observer.modules[0].found == false);
+}
+
+// ======================================================================================
+
+KJ_TEST("A user bundle with a single ESM module") {
+  ModuleBundle::BundleBuilder builder;
+
+  kj::String source = kj::str("export const foo = 123;");
+  builder.addEsmModule("foo", source.releaseArray(), Module::Flags::MAIN);
+
+  auto bundle = builder.finish();
+
+  const auto specifier = "file:///foo"_url;
+
+  // Resolve should return nothing.
```

Comment:
Resolve does return the ESM module here.

Reply:
Yeah the comment is out of date
```suggestion
```